refactor(compiler): required inputs prerequisite refactors

crisbeto opened the pull request:
Based on the discussion in https://github.com/angular/angular/pull/49304#discussion_r1124732608. Reworks the compiler internals to allow for additional information about inputs to be stored. This is a prerequisite for required inputs.

crisbeto commented on packages/compiler-cli/src/ngtsc/metadata/src/property_mapping.ts:
```diff
@@ -23,49 +23,103 @@ export type ClassPropertyName = string;
 
 /**
- * An input or output of a directive that has both a named JavaScript class property on a component
+ * A binding of a directive that has both a named JavaScript class property on a component
```
The diff in this file is pretty large, but the behavior is the same as before. I just split up the input-specific and output-specific code from `ClassPropertyMapping` into two separate classes so that inputs can be marked as required in the next PR.

crisbeto commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -89,7 +89,7 @@ export interface R3DirectiveMetadata {
    * binding property name and class property name if the names are different.
    */
-  inputs: {[field: string]: string|[string, string]};
+  inputs: {[field: string]: {alias: string, classPropertyName: string}|string};
```
Note here: technically we should be able to remove the `|string` for the values, but the way `ngUpgrade` maps inputs between AngularJS and Angular components depends specifically on the `string` signature further downstream. I decided to keep it for now, because I want to avoid large refactors in `ngUpgrade` and because the string signature doesn't require much code to support.

alxhub replied:
I don't think we should change the format of generated code at all - the current `string|[string, string]` is optimized for code size. Switching to an object literal with named fields definitely makes it more readable, but is likely a significant size regression.

crisbeto replied:
FWIW this isn't the generated code, it's just the common interface between AOT and JIT. The generated code is the same as before.

crisbeto commented on packages/core/test/render3/jit/declare_component_spec.ts:
```diff
@@ -525,23 +525,25 @@ function expectComponentDef(
-  expect(actual.onPush).toEqual(expectation.onPush);
-  expect(actual.styles).toEqual(expectation.styles);
-  expect(actual.data).toEqual(expectation.data);
+  expect(actual.selectors).withContext('selectors').toEqual(expectation.selectors);
```
I added all of these `withContext` calls to make the failures easier to read.

crisbeto commented:
[Passing TGP](http://test/OCL:514674933:BASE:514760641:1678211173388:60758be7).

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -512,76 +510,140 @@ function isPropertyTypeMember(member: ClassMember): boolean {
+        inputsField, inputsArray, `Failed to resolve @Directive.inputs to a string array`);
+  }
+
+  for (let i = 0; i < inputsArray.length; i++) {
+    const value = inputsArray[i];
```
Will `i` end up being used besides reading from `inputsArray` with required inputs? If not, this could become a `for-of` loop

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -512,76 +510,140 @@ function isPropertyTypeMember(member: ClassMember): boolean {
-    // be undefined, in which case the field name should also be used as the property name.
-    const [field, property] = value.split(':', 2).map(str => str.trim());
-    results[field] = property || field;
+    const [publicName, fieldName] = parseMappingString(value);
```
I see both `public` and `alias` being used now, would it make sense to always switch to `alias`?

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/property_mapping.ts:
```diff
@@ -98,59 +152,76 @@ export class ClassPropertyMapping implements InputOutputPropertySet {
-  }
+/** Variation of `ClassPropertyMapping` that include information specific to inputs. */
+export class ClassInputPropertyMapping extends BasePropertyMapping {
+  private requiredBindingNames = new Set<BindingPropertyName>();
```
Did you intent to include this in this PR?

crisbeto replied:
I left it here for myself since this ties into the required inputs PR.

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/property_mapping.ts:
```diff
@@ -23,49 +23,103 @@ export type ClassPropertyName = string;
    * The property name used to bind this input or output in an Angular template.
    */
-  readonly bindingPropertyName: BindingPropertyName;
+  readonly alias: BindingPropertyName;
```
We're still all-over-the-place with naming. Since this PR is already large I wouldn't want to introduce additional renames, but it wouldn't hurt to go over this in another pass to become more consistent with naming.

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/property_mapping.ts:
```diff
@@ -23,49 +23,103 @@ export type ClassPropertyName = string;
+  readonly required?: boolean;  // TODO(required-inputs): make non-optional
+}
+
+abstract class BasePropertyMapping implements InputOutputPropertySet {
```
I'm not sure I like this split/base class setup, as it seems simpler to have something like the following:

```ts
class ClassPropertyMapping<P extends PropertyMapping> implements InputOutputPropertySet {}
```

Methods like `hasRequiredBindingName` can then no longer exist on this class, and have to be implemented using `getByBindingPropertyName().some(mapping => mapping.required)`, but I'd argue this is desirable: the property mapping's concern is about providing a two-way association of alias/field names and their metadata, _not_ interpreting that metadata.

I might be missing something here.

crisbeto replied:
I agree that this is a little janky, but I wanted to avoid reusing the exact same data structure both for inputs and outputs since outputs are likely to stay as they are while we'll probably add some more properties for the inputs. The problem with your approach is that the mapping gets constructed from a `Record<string, string|[string, string]>` so there's no place to encode the additional field.

JoostK replied:
Yeah, that would need to become `Record<string, P>`, where `<P extends PropertyMapping>`. I'm not sure how that pans out for real.

alxhub replied:
+1 to this suggestion - I like the idea of a generic data structure for these property mappings, which allows for additional metadata to be present on the metadata it stores.

JoostK commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -661,12 +664,41 @@ function isOutput(value: any): value is Output {
-function parseInputOutputs(values: string[]): StringMap {
+function inputsMappingToInputMetadata(inputs: Record<string, string|[string, string]>) {
+  return Object.keys(inputs).reduce((result, key) => {
+    const value = inputs![key];
```
```suggestion
    const value = inputs[key];
```

JoostK commented on packages/compiler/src/compiler_facade_interface.ts:
```diff
@@ -165,7 +163,7 @@ export interface R3DirectiveMetadataFacade {
   propMetadata: {[key: string]: OpaqueValue[]};
   lifecycle: {usesOnChanges: boolean;};
-  inputs: string[];
+  inputs: (string|{name: string, alias?: string})[];  // TODO(required-inputs): add required flag
```
Can we normalise this to just the object-literal type? I don't quite see why we retain the bare `string` type here.

crisbeto replied:
That's what I was referring to here: https://github.com/angular/angular/pull/49333#discussion_r1128017957.

JoostK replied:
I already thought so, but that comment is for code in the compiler, whereas this is the JIT bridge to the compiler which I suppose is not concerned with `ngUpgrade` usages?

JoostK commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -423,12 +423,20 @@ export function createBaseDirectiveTypeParams(meta: R3DirectiveMetadata): o.Type
+  // TODO(required-inputs): expand this to generate the new object literal syntax.
+  return o.literalMap(Object.keys(meta.inputs).map(key => {
+    const value = meta.inputs[key];
+    return ({key, value: o.literal(typeof value === 'string' ? value : value.alias), quoted: true});
```
This appears to have been pulled from an arrow function's expression body, but now that it's a return statement the parens are redundant:

```suggestion
    return {key, value: o.literal(typeof value === 'string' ? value : value.alias), quoted: true};
```

crisbeto commented on packages/compiler-cli/src/ngtsc/metadata/src/dts.ts:
```diff
@@ -167,6 +167,29 @@ export class DtsMetadataReader implements MetadataReader {
   }
 }
 
+function readInputsType(type: ts.TypeNode): {[field: string]: InputPropertyMapping} {
```
This is currently a placeholder, but it'll have some more input-specific logic in the next PR.

crisbeto committed:
refactor(compiler): required inputs prerequisite refactors

Based on the discussion in https://github.com/angular/angular/pull/49304#discussion_r1124732608. Reworks the compiler internals to allow for additional information about inputs to be stored. This is a prerequisite for required inputs.

crisbeto commented:
@JoostK @alxhub I've reworked the PR based on your feedback. Can you take another look?

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/shared.ts:
```diff
@@ -512,76 +509,142 @@ function isPropertyTypeMember(member: ClassMember): boolean {
+    }
+
+    inputs[fieldName] = {
+      bindingPropertyName: bindingPropertyName ?? fieldName,
```
Is this fallback necessary? `bindingPropertyName` is non-nullable, so something appears off here.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/symbol.ts:
```diff
@@ -87,10 +87,13 @@ export class DirectiveSymbol extends SemanticSymbol {
-    previous: [ClassPropertyName, BindingPropertyName]): boolean {
-  return current[0] === previous[0] && current[1] === previous[1];
+function isInputMappingEqual(current: InputMapping, previous: InputMapping): boolean {
+  return isInputOrOutputEqual(current, previous) && current.required === previous.required;
```
For a future PR where required inputs are added, we should add a test for this in `packages/compiler-cli/test/ngtsc/incremental_semantic_changes_spec.ts`.

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/property_mapping.ts:
```diff
@@ -153,24 +163,18 @@ export class ClassPropertyMapping implements InputOutputPropertySet {
     return obj;
   }
 
+  // TODO: rename this method?
```
Is this pending? I don't really have suggestions for an alternative name here.

JoostK commented on packages/compiler-cli/src/ngtsc/metadata/src/property_mapping.ts:
```diff
@@ -179,16 +183,16 @@ export class ClassPropertyMapping implements InputOutputPropertySet {
-    for (const [classPropertyName, inputOrOutput] of this.forwardMap.entries()) {
-      yield [classPropertyName, inputOrOutput.bindingPropertyName];
+  * [Symbol.iterator](): IterableIterator<T> {
+    for (const [_classPropertyName, inputOrOutput] of this.forwardMap.entries()) {
```
Consider iterating over `.values()` to avoid needing to destructure and ignoring the key.

JoostK commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -661,12 +664,45 @@ function isOutput(value: any): value is Output {
+function parseInputsArray(values: (string|{name: string, alias?: string})[]) {
+  return values.reduce((results, value) => {
+    if (typeof value === 'string') {
+      const [bindingPropertyName, fieldName] = parseMappingString(value);
```
Rename `fieldName` to `classPropertyName` here to use shorthand property assignment in the line below.

JoostK reviewed (approved):
I like the cleanup around `ClassPropertyMapping`! Some nits for your consideration but otherwise LGTM

alxhub commented on packages/compiler/src/render3/view/util.ts:
```diff
@@ -163,31 +163,32 @@ export function asLiteral(value: any): o.Expression {
-      needsDeclaredName = publicName !== declaredName;
-    } else {
+    if (typeof value === 'string') {
       minifiedName = declaredName = key;
```
Nit: I think it's more readable to not use assignment chaining (if we aren't optimizing for conciseness / code size).

alxhub commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -85,11 +85,12 @@ export interface R3DirectiveMetadata {
    * binding property name and class property name if the names are different.
    */
-  inputs: {[field: string]: string|[string, string]};
+  inputs: {[field: string]: {classPropertyName: string, bindingPropertyName: string}};
```
I would suggest creating the interface:

```typescript
export interface R3InputMetadata {
  classPropertyName: string;
  bindingPropertyName: string;
}
```

Then the `inputs` declaration could become:
```suggestion
  inputs: R3InputMetadata[],
```

Using an array instead of an object also makes the downstream iteration easier, and avoids the need to do `Object.key` operations to iterate it.

alxhub commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -423,12 +423,24 @@ export function createBaseDirectiveTypeParams(meta: R3DirectiveMetadata): o.Type
+    const value = meta.inputs[key];
+    return {
+      key,
+      value: o.literal(typeof value === 'string' ? value : value.bindingPropertyName),
```
Can `typeof value` be a `string` here? We changed the type to always be always be an object (instance of `R3InputMetadata`).

alxhub commented on packages/compiler/src/render3/view/util.ts:
```diff
@@ -163,31 +163,32 @@ export function asLiteral(value: any): o.Expression {
-    keys: {[key: string]: string|string[]}, keepDeclared?: boolean): o.Expression|null {
-  if (Object.getOwnPropertyNames(keys).length > 0) {
-    return mapToExpression(keys, keepDeclared);
+export function conditionallyCreateMapObjectLiteral(map: Record<string, string|{
```
Nit: we should name this something to do with inputs/outputs, for clarity.

crisbeto committed:
fixup! refactor(compiler): required inputs prerequisite refactors

crisbeto commented:
[Almost passing TGP](http://test/OCL:514674933:BASE:516481213:1678791604448:ffb850b9). There's one unrelated failure.

alxhub commented:
This PR was merged into the repository by commit be97c8702319c9e1467c4fddba868e8f23ebe159.
